scripts/bootloader/generate_zip.py

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+
+import argparse
+import json
+import time
+from zipfile import ZipFile
+from os import path, remove
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="Generate zip file for binary artifact. Provide a list of 'key=value' pairs separated by space for"
+                    "information that should be stored in manifest.json. To provide information specific for one of "
+                    "the files, prepend the 'key' with the name of the file (only the basename, not the path)",
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+
+    parser.add_argument("--output", required=True, type=argparse.FileType(mode='w'), help="Output zip path")
+    parser.add_argument("--bin-files", required=True, type=argparse.FileType(mode='r'), nargs='+',
+                        help="Bin files to be stored in zip")
+    return parser.parse_known_args()
+
+
+def get_name(text_wrapper):
+    return path.basename(text_wrapper.name)
+
+
+if __name__ == "__main__":
+    args, info = parse_args()
+
+    manifest = dict()
+
+    if path.exists(args.output.name):
+        remove(args.output.name)
+
+    manifest['format-version'] = 0
+    manifest['time'] = int(time.time())
+    manifest['files'] = list()
+
+    shared_info = dict()
+    special_info = dict()
+    name_to_path = dict()
+
+    for f in args.bin_files:
+        name_to_path[get_name(f)] = f.name
+
+    for i in info:
+        special = False
+        key = i.split('=')[0]
+        val = i.split('=')[1]
+        if val.startswith('0x'):
+            val = int(val, base=16)
+
+        # When multiple bin files are given, the non-shared configurations are prepended with the name of the bin file
+        for p in name_to_path.keys():
+            if p not in special_info:
+                special_info[p] = dict()
+            if key.startswith(p):
+                special_info[p][key.replace(p, '')] = val
+                special = True
+        if not special:
+            shared_info[i.split('=')[0]] = val
+
+    for n, p in name_to_path.items():
+        special_info[n]['size'] = path.getsize(p)
+        special_info[n]['file'] = n
+        special_info[n]['modtime'] = int(path.getmtime(p))
+        merged = dict()
+        merged.update(special_info[n])
+        merged.update(shared_info)
+        manifest['files'].append(merged)
+
+    with ZipFile(args.output.name, 'w') as my_zip:
+        my_zip.writestr('manifest.json', json.dumps(manifest, indent=4))
+        for n, p in name_to_path.items():
+            my_zip.write(p, n)
```

Comment:
Unless there is a measurable performance issue with creating the .json file, I think the json file should exist as a file in the build directory.

This will make debugging easier. E.g. grepping for format-version in the build dir will show the json file.

Reply:
Agree! I use the file name `manifest.json_${NAME_OF_DFU_OUTPUT}` to ensure that all zip manifests are available and ack/grep/find'able.